metrics/src/metadata.rs

```diff
@@ -0,0 +1,46 @@
+#![allow(missing_docs)]
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Level(LevelInner);
+
+impl Level {
+    pub const TRACE: Self = Self(LevelInner::Trace);
+    pub const DEBUG: Self = Self(LevelInner::Debug);
+    pub const ERROR: Self = Self(LevelInner::Error);
+    pub const WARN: Self = Self(LevelInner::Warn);
+    pub const INFO: Self = Self(LevelInner::Info);
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum LevelInner {
+    Trace = 0,
+    Debug = 1,
+    Info = 2,
+    Warn = 3,
+    Error = 4,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Metadata<'a> {
+    target: &'a str,
+    level: Level,
+    module_path: Option<&'a str>,
```

Comment:
Hmm, why exactly is module path optional but not target?

Seems like we could swap that around, and then have `target()` return `self.target.as_ref().unwrap_or(self.module_path)` or something like that.

Reply:
I've just imitated `tracing`s behavior here.

https://github.com/tokio-rs/tracing/blob/e603c2a254d157a25a7a1fbfd4da46ad7e05f555/tracing-core/src/metadata.rs#L68-L70

> The name of the Rust module where the span occurred, or `None` if this could not be determined.

Admittedly I don't know _when_ this cannot be determined given that `module_path!` returns a `&'static str`.

I will do some investigation about the reason why `tracing::Metadata` is setup like this and then get back to you.

Reply:
Summarizing my brief discussion with the `tracing` maintainers:

- `module_path` should _always_ be populated when the `Metadata` is constructed from the `tracing` macros.
- It is represented as an `Option`:
    1. To maintain compatibility with `log` whose [`module_path`](https://docs.rs/log/latest/log/struct.Record.html#method.module_path) is also optional.
    1. To interop with events emitted in contexts where `module_path` might not be readily accessible (outside of Rust).

We don't need to be compatible with `log` however it might be nice to enable the use case (ii).